bin/propolis-server/src/lib/migrate/source.rs

```diff
@@ -60,6 +60,18 @@ struct SourceProtocol<T: AsyncRead + AsyncWrite + Unpin + Send> {
 
     /// Transport to the destination Instance.
     conn: WebSocketStream<T>,
+
+    // TODO: A hacky way for now to hang onto the VMM timing data between
+    // migration phases.
+    //
+    // We want to read the VMM timing data as soon as we can after we pause the
+    // source, and make adjustments on the destination as close as we can to
+    // the end of the migration.
+    //
+    // This lets us hang on to the data between export in the pause phase,
```

Comment:
> The destination has to compute the total amount of time the VM spent paused in any case

I'm not sure it so much has to calculate how much time the VM was paused, but rather how much time passed between when the reading of timing (TSC, etc) information was taken, and when its adjusted value(s) are being loaded on the destination.  The destination VM could remain paused for additional time past that, but it would appear no different than a case where vCPU threads are not being scheduled for some reason.

Contextually, it is important that the timing information be loaded first, before other device state, since the adjusted `boot_hrtime` value is used to offset the other timer values in the emulated device state.

Reply:
> How deeply do we care about which protocol phase sends this data?

As Patrick said, it needs to be sent before the device state is imported (which this current change does *not* do), but also we are trying to push most of the adjustments made into userspace, which implies we need to collect the data before any lengthier pieces of migration, where possible. Perhaps I was a bit too constraining when describing this in terms of when we pause the vCPU, but it's been my mental model for thinking about it. I can change some of the comments here to reflect that.

I had avoided reintroducing the ArchState phase in the hopes that it would be cleaner given that we split up the collecting of time data and the sending of it. But perhaps that isn't true, as it might be harder to grok that this data must be imported before the device state is imported.

I could add the ArchState phase back as such: Pause (with time data collection) -> RamPush -> ArchState (time data send) -> DeviceState...

How does that sound?

Reply:
In 803081, I added:
- a TimeDataRead step on the source side
- a TimeData phase for both sides to exchange the data

Reply:
see: https://github.com/oxidecomputer/propolis/pull/359#discussion_r1191799079

I consolidated down to one protocol phase